lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/MFMA.cpp

```diff
@@ -331,13 +334,29 @@ struct DotOpMFMAConversionHelper {
     return success();
   }
 
-  ValueTable getValuesFromDotOperandLayoutStruct(Value value, int n0, int n1,
+/**
+ * @brief Converts dot operand structure to value table and converts types appropriate for mfma instructions
+*/
+  ValueTable getValuesFromDotOperandLayoutStruct(Value value, int n0, int n1, int kWidth,
                                                  Type type) const {
     auto elems = typeConverter->unpackLLElements(loc, value, rewriter, type);
     ValueTable vals;
     for (int i = 0; i < n0; i++) {
       for (int j = 0; j < n1; j++) {
-        vals[{i, j}] = elems[n1 * i + j];
+        auto rawElems = elems[n1 * i + j];
+        Value convertedElems;
+        if (type.isF32()) {
+          convertedElems = extract_element(type, rawElems, i32_val(0));
```

Comment:
Why do we need to dereference rawElems here for i32 input?

Reply:
This part is related to processing of A/B dot operands.
This code is an adapter from generic `vec<base_type x kwidth>` format to the format that rocdl intrinsic expects.

Previously this transformation was done in Type converter (see [TypeConverter.cpp](https://github.com/ROCmSoftwarePlatform/triton/pull/400/files#diff-928f898f93db4afa3551c43a4bdc36b127c0c866f76ac7cdc90bae10116972e9) below), but I feel that it is better to have this transformation closer to MFMA emitting code. So I simplified type converter and moved this code here.

A and B could be one of variety of types: fp32, fp16, bf16, int8, fp8*
But rocdl mfma intrinsics takes some of these types in a packed integer format.

For example:
- `fp32xfp32 -> fp32` version intrinsic takes plain scalar `fp32`, `fp32` as A/B arguments.
- `fp8xfp8 -> fp32` or `int8xint8 -> fp32` version takes several values in for of packed int32 or int64 (depending on kwidth of operation)